android: log cleared pending exceptions to platform layer

Augustyniak opened the pull request:
Commit Message: The main purpose of this PR is to add logging of the information about pending exceptions found by JNI layer to the platform layer. That should help us to understand better cases where the library crashes/misbehaves due to Java exception being thrown and not handled correctly by our JNI layer. 
Additional Description: I made a little bit of a refactor as I moved methods related to `JNIEnv` to their own c++ `Envoy::JNI::Env` class and methods related to `JavaVM` to their own `JavaVirtualMachine` c++ class. That helped me to introduce an `Envoy::JNI::String` class that makes working with `jstring` type easier and less error-prone (hopefully).
Risk Level: Low
Testing: Integration
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

alyssawilk commented:
/wait
@Augustyniak if this is WIP mind marking it as a draft so it doesn't show up in the maintainer oncall reminders?

Augustyniak commented:
@danzh2010 after this is merged we can add `Envoy::JNI::Exception::checkAndClear` calls to platform cert verification code and it should help us to get more details about any exceptions that are thrown as part of our platform cert validation logic.

Augustyniak commented on mobile/library/common/jni/jni_utility.cc:
```diff
@@ -38,23 +32,7 @@ jclass find_class(const char* class_name) {
   return clazz;
 }
 
-JNIEnv* get_env() {
```
This is being moved to Envoy::JNI:: Env c++ class. No functional changes

Augustyniak commented on mobile/library/common/jni/jni_utility.cc:
```diff
@@ -6,17 +6,11 @@
 static jobject static_class_loader = nullptr;
-static thread_local JNIEnv* local_env = nullptr;
-
-void set_vm(JavaVM* vm) { static_jvm = vm; }
```
This is being moved to Envoy::JNI::JavaVirtualMachine c++ class. No functional changes

snowp commented on mobile/library/common/jni/types/JavaVirtualMachine.h:
```diff
@@ -0,0 +1,27 @@
+  static void initialize(JavaVM* jvm) { jvm_ = jvm; }
+  static JavaVM* getJavaVM() { return jvm_; }
+  static jint getJNIVersion();
+  static void detachCurrentThread() { jvm_->DetachCurrentThread(); }
```
Maybe ASSERT on the return value? 

In general it feels like it would be more C++ to have this be backed by a thread local object that was responsible for calling detach on deinit, that way nothing is responsible for manually calling this

Augustyniak replied:
Added assert.

snowp commented on mobile/library/common/jni/types/Exception.cc:
```diff
@@ -0,0 +1,154 @@
+ * GENERIC_EXCEPTION_DESCRIPTION||EXCEPTION_STACKTRACE||EXCEPTION_CAUSE_STACKTRACE
+ */
+std::string Exception::description() const {
+  std::vector<std::string> comps = {
```
avoid abbreviations in variable names

snowp commented on mobile/library/common/jni/jni_interface.cc:
```diff
@@ -242,7 +244,7 @@ static void* jvm_on_headers(const char* method, const Envoy::Types::ManagedEnvoy
                                          end_stream ? JNI_TRUE : JNI_FALSE, j_stream_intel);
   // TODO(Augustyniak): Pass the name of the filter in here so that we can instrument the origin of
   // the JNI exception better.
```
What I've done in similar situation is to include a `detail` param that is some arbitrary string that describes the context in which this exception is being handled in

Augustyniak replied:
I like the idea - implemented

Augustyniak replied:
implemented ur suggestion 👍

snowp reviewed (changes requested):
Nice, a few comments

Augustyniak commented on mobile/library/common/jni/types/String.h:
```diff
@@ -0,0 +1,47 @@
+   * in modified UTF-8 encoding. The view is valid for as long as the String wrapping it stays
+   * alive.
+   */
+  absl::string_view get() { return absl::string_view(string_); }
```
That being said, I think that I need to handle nullptr `string_` in a special way - updating.

snowp commented on mobile/library/common/jni/types/Env.cc:
```diff
@@ -0,0 +1,32 @@
+  return local_env_;
+}
+
+thread_local JNIEnv* Env::local_env_ = nullptr;
```
For this to detach on exit you need some RAII wrapper that calls detach in its dtor, similar to this https://github.com/jni-rs/jni-rs/blob/master/src/wrapper/java_vm/vm.rs#L523

Augustyniak replied:
that thread_local is for `JNIEnv` - we use `static` for storing `JavaVM`.

Augustyniak replied:
nvm - let my try something

Augustyniak replied:
OK, I did attempt to go down this road but quickly ran into issues with the initialization of global variables / thread-safe in here:

```
[mutex.cc : 1404] RAW: Potential Mutex deadlock: 
        @ 0x2d8f14dd0 absl::DebugOnlyDeadlockCheck()
        @ 0x2d8f14f58 absl::Mutex::ReaderLock()
        @ 0x2d634cd1c absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d634cb78 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d86c60b0 Envoy::Logger::DelegatingLogSink::log()
        @ 0x2d86bd970 spdlog::logger::sink_it_()
        @ 0x2d3c02328 spdlog::logger::log_it_()
        @ 0x2d81ba154 spdlog::logger::log_<>()
        @ 0x2d81b9eac spdlog::logger::log<>()
        @ 0x2d81b7c48 Envoy::BackwardsTrace::logFault()
        @ 0x2d81b7698 Envoy::SignalAction::sigHandler()
        @ 0x1873602a4 _sigtramp
        @ 0x18707da00 abort_with_payload_wrapper_internal
        @ 0x18707da34 dyld
        @ 0x18700c0a4 dyld4::CacheFinder::~CacheFinder()
        @ 0x18704027c dyld4::APIs::_tlv_atexit()
        @ 0x2d70115e4 __cxx_global_var_init
        @ 0x2d7011560 __tls_init
        @ 0x2d700f078 _ZTWN5Envoy3JNI3Env10local_env_E
        @ 0x2d700ec24 Envoy::JNI::Env::get()
```

 I say we move forward with what we have in here now and revisit this later on - I am at the point where I do not want to spend more time fighting c++ in here. If anything, I would prefer to address `call_method` comment before I fight thread_safe initialization.

Augustyniak commented on mobile/library/common/jni/types/Exception.cc:
```diff
@@ -0,0 +1,154 @@
+  return std::string(String(j_description).get());
+}
+
+std::string Exception::throwableStacktraceDescription(jthrowable throwable) const {
```
Unless you have a strong opinion I would prefer to keep it the way it is + introduce more JNI helpers to make code easier to follow.

Augustyniak commented:
/retest

snowp reviewed (changes requested):
the C++ files should be lower cased, otherwise lgtm

alyssawilk commented on mobile/library/common/jni/jni_interface.cc:
```diff
@@ -20,12 +21,13 @@ using Envoy::Platform::EngineBuilder;
 JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM* vm, void* reserved) {
   JNIEnv* env = nullptr;
-  if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION) != JNI_OK) {
+  if (vm->GetEnv(reinterpret_cast<void**>(&env), Envoy::JNI::JavaVirtualMachine::getJNIVersion()) !=
```
optional nit but if this is the only use case it might be easier to make this isJniVersionOk just to keep the code simpler

Augustyniak replied:
updated per your suggestion

alyssawilk commented on mobile/library/common/jni/types/exception.cc:
```diff
@@ -0,0 +1,152 @@
+      throwableDescription(throwable_),
+      detail == "" ? "NO_DETAIL" : detail,
+      throwableStacktraceDescription(throwable_),
+      throwable_cause_description == "" ? "NO_CAUSE" : throwable_cause_description, 
```
is NO_CAUSE accurate or is it more UNKNOWN_CAUSE?

Augustyniak replied:
I found some docs about the `cause` property:

>getCause - Returns the cause of this throwable or null if the cause is nonexistent or unknown. (The cause is the throwable that caused this throwable to get thrown.)

so I think that I will either go with your suggestion `CAUSE_UNKNOWN` or `CAUSE_UNKNOWN_OR_NONEXISTENT`

alyssawilk commented on mobile/library/common/jni/types/exception.h:
```diff
@@ -0,0 +1,40 @@
+public:
+  /**
+   * @brief Checks and clears any pending exceptions. Reports pending exceptions to a platform
+   * layer.
```
worth documenting how it's reported?  AFIK really it just logs a message which may or may not be intercepted at the platform layer

Augustyniak replied:
added more details

alyssawilk commented on mobile/test/kotlin/integration/FilterThrowingExceptionTest.kt:
```diff
@@ -95,18 +95,26 @@ class FilterThrowingExceptionTest {
+    val onJNIExceptionEventLatch = CountDownLatch(2)
+
+    var expectedMessages = mutableListOf(
+      "Simulated onRequestHeaders exception",
```
This test makes me so happy :-)

alyssawilk reviewed (commented):
Very cool!  Looks like you need format check

Augustyniak commented:
The good thing is that I think that I addressed code review comments. The bad thing is that I ran into the following crash (happened only once):
```
[2023-03-22 15:38:15.231][12864754][info][misc] [library/common/jni/types/exception.cc:20] java.lang.Exception: Simulated onRequestHeaders exception||onRequestHeaders||test.kotlin.integration.ThrowingFilter.onRequestHeaders(FilterThrowingExceptionTest.kt:42);io.envoyproxy.envoymobile.EnvoyHTTPFilterAdapter.onRequestHeaders(Filter.kt:33);io.envoyproxy.envoymobile.engine.JvmFilterContext.onRequestHeaders(JvmFilterContext.java:57)||CAUSE_UNKNOWN_OR_NONEXISTENT
WARNING in native method: JNI call made without checking exceptions when required to from CallObjectMethodV
WARNING in native method: JNI call made without checking exceptions when required to from CallObjectMethodV
WARNING in native method: JNI call made without checking exceptions when required to from CallObjectMethodV
[mutex.cc : 1404] RAW: Potential Mutex deadlock: 
        @ 0x2d8f83af0 absl::DebugOnlyDeadlockCheck()
        @ 0x2d8f83c78 absl::Mutex::ReaderLock()
        @ 0x2d63af978 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d63af7d4 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d8734e08 Envoy::Logger::DelegatingLogSink::log()
        @ 0x2d872c6c8 spdlog::logger::sink_it_()
        @ 0x2d3c61348 spdlog::logger::log_it_()
        @ 0x2d8228dcc spdlog::logger::log_<>()
        @ 0x2d8228b24 spdlog::logger::log<>()
        @ 0x2d82268c0 Envoy::BackwardsTrace::logFault()
        @ 0x2d8226310 Envoy::SignalAction::sigHandler()
        @ 0x1873602a4 _sigtramp
        @ 0x10eea0ad8 0x0

[mutex.cc : 1416] RAW: Acquiring absl::Mutex 0x600003fc30d8 while holding  0x600003fc30d8; a cycle in the historical lock ordering graph has been observed
[mutex.cc : 1417] RAW: Cycle: 
[mutex.cc : 1431] RAW: mutex@0x600003fc30d8 stack: 
        @ 0x2d8f83af0 absl::DebugOnlyDeadlockCheck()
        @ 0x2d8f83c78 absl::Mutex::ReaderLock()
        @ 0x2d63af978 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d63af7d4 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d8734704 Envoy::Logger::DelegatingLogSink::delegate()
        @ 0x2d87346b8 Envoy::Logger::SinkDelegate::setDelegate()
        @ 0x2d87348d8 Envoy::Logger::StderrSinkDelegate::StderrSinkDelegate()
        @ 0x2d873492c Envoy::Logger::StderrSinkDelegate::StderrSinkDelegate()
        @ 0x2d8735458 std::__1::make_unique<>()
        @ 0x2d8735308 Envoy::Logger::DelegatingLogSink::init()
        @ 0x2d63a4c3c Envoy::Logger::Registry::getSink()
        @ 0x2d5356ec4 Envoy::Engine::main()
        @ 0x2d53696e0 _ZNSt3__1L8__invokeIMN5Envoy6EngineEF14envoy_status_tONS_10unique_ptrINS1_11OptionsImplENS_14default_deleteIS5_EEEEEPS2_JS8_EvEEDTcldsdescT0_fp0_fp_spscT1_fp1_EEOT_OSD_DpOSE_
        @ 0x2d5369620 std::__1::__thread_execute<>()
        @ 0x2d5368af0 std::__1::__thread_proxy<>()
        @ 0x18733206c _pthread_start
        @ 0x18732ce2c thread_start

[mutex.cc : 1436] RAW: dying due to potential deadlock
[mutex.cc : 1404] RAW: Potential Mutex deadlock: 
        @ 0x2d8f83af0 absl::DebugOnlyDeadlockCheck()
        @ 0x2d8f83c78 absl::Mutex::ReaderLock()
        @ 0x2d63af978 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d63af7d4 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d8734e08 Envoy::Logger::DelegatingLogSink::log()
        @ 0x2d872c6c8 spdlog::logger::sink_it_()
        @ 0x2d3c61348 spdlog::logger::log_it_()
        @ 0x2d8228dcc spdlog::logger::log_<>()
        @ 0x2d8228b24 spdlog::logger::log<>()
        @ 0x2d82268c0 Envoy::BackwardsTrace::logFault()
        @ 0x2d8226310 Envoy::SignalAction::sigHandler()
        @ 0x1873602a4 _sigtramp
        @ 0x187331cec pthread_kill
        @ 0x18726a2c8 abort
        @ 0x2d8ff45ec absl::raw_log_internal::(anonymous namespace)::RawLogVA()
        @ 0x2d8ff4458 absl::raw_log_internal::RawLog()
        @ 0x2d8f895d8 absl::DeadlockCheck()
        @ 0x2d8f83af0 absl::DebugOnlyDeadlockCheck()
        @ 0x2d8f83c78 absl::Mutex::ReaderLock()
        @ 0x2d63af978 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d63af7d4 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d8734e08 Envoy::Logger::DelegatingLogSink::log()
        @ 0x2d872c6c8 spdlog::logger::sink_it_()
        @ 0x2d3c61348 spdlog::logger::log_it_()
        @ 0x2d8228dcc spdlog::logger::log_<>()
        @ 0x2d8228b24 spdlog::logger::log<>()
        @ 0x2d82268c0 Envoy::BackwardsTrace::logFault()
        @ 0x2d8226310 Envoy::SignalAction::sigHandler()
        @ 0x1873602a4 _sigtramp
        @ 0x10eea0ad8 0x0

[mutex.cc : 1416] RAW: Acquiring absl::Mutex 0x600003fc30d8 while holding  0x600003fc30d8; a cycle in the historical lock ordering graph has been observed
[mutex.cc : 1417] RAW: Cycle: 
[mutex.cc : 1431] RAW: mutex@0x600003fc30d8 stack: 
        @ 0x2d8f83af0 absl::DebugOnlyDeadlockCheck()
        @ 0x2d8f83c78 absl::Mutex::ReaderLock()
        @ 0x2d63af978 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d63af7d4 absl::ReaderMutexLock::ReaderMutexLock()
        @ 0x2d8734704 Envoy::Logger::DelegatingLogSink::delegate()
        @ 0x2d87346b8 Envoy::Logger::SinkDelegate::setDelegate()
        @ 0x2d87348d8 Envoy::Logger::StderrSinkDelegate::StderrSinkDelegate()
        @ 0x2d873492c Envoy::Logger::StderrSinkDelegate::StderrSinkDelegate()
        @ 0x2d8735458 std::__1::make_unique<>()
        @ 0x2d8735308 Envoy::Logger::DelegatingLogSink::init()
        @ 0x2d63a4c3c Envoy::Logger::Registry::getSink()
        @ 0x2d5356ec4 Envoy::Engine::main()
        @ 0x2d53696e0 _ZNSt3__1L8__invokeIMN5Envoy6EngineEF14envoy_status_tONS_10unique_ptrINS1_11OptionsImplENS_14default_deleteIS5_EEEEEPS2_JS8_EvEEDTcldsdescT0_fp0_fp_spscT1_fp1_EEOT_OSD_DpOSE_
        @ 0x2d5369620 std::__1::__thread_execute<>()
        @ 0x2d5368af0 std::__1::__thread_proxy<>()
        @ 0x18733206c _pthread_start
        @ 0x18732ce2c thread_start

[mutex.cc : 1436] RAW: dying due to potential deadlock
```

Augustyniak commented:
I was not able to repro the aforementioned crash. Among others, I tried `./bazelw test --runs_per_test=100 --build_tests_only //test/kotlin/integration:filter_throwing_exception_test`

alyssawilk reviewed (approved):
LGTM but can you run the kotlin test at least 2-3k times?  that's my standard "will not constantly annoy us on CI" flake threshold.

Augustyniak commented:
Ran `./bazelw test --runs_per_test=2000 --build_tests_only //test/kotlin/integration:filter_throwing_exception_test`

in my PR:
21 failures

on main branch:
40 failures

There results are not great by any means but at least gives me some confidence that changes from my PR are not making things worse.
